File: backend/app/schemas/sbom.py
"""Schemas Pydantic des SBOM (méta + composants)."""
from datetime import datetime

from pydantic import BaseModel, Field


class SbomComponentResponse(BaseModel):
    purl: str | None
    name: str
    version: str | None
    component_type: str | None

    model_config = {"from_attributes": True}


class SbomResponse(BaseModel):
    """Méta d'un SBOM importé (+ warnings du dernier parsing à l'upload)."""

    format: str
    spec_version: str
    filename: str | None
    component_count: int
    imported_at: datetime
    warnings: list[str] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class SbomDetailResponse(SbomResponse):
    """Méta + page de composants."""

    components: list[SbomComponentResponse] = Field(default_factory=list)
    total_components: int = 0


class SbomSummaryItem(BaseModel):
    """Entrée du résumé SBOM d'un arbre (assets ayant un SBOM)."""

    asset_id: str
    format: str
    component_count: int
    imported_at: datetime

File: frontend/src/api/sbom.ts
import { api } from './client';

export interface SbomComponent {
  purl: string | null;
  name: string;
  version: string | null;
  component_type: string | null;
}

export interface SbomMeta {
  format: string;
  spec_version: string;
  filename: string | null;
  component_count: number;
  imported_at: string;
  warnings: string[];
}

export interface SbomDetail extends SbomMeta {
  components: SbomComponent[];
  total_components: number;
}

export interface SbomSummaryItem {
  asset_id: string;
  format: string;
  component_count: number;
  imported_at: string;
}

export const sbomApi = {
  getSummary: (treeId: number) =>
    api.get<SbomSummaryItem[]>(`/assets/sbom/summary?tree_id=${treeId}`),

  getSbom: (assetId: string, treeId: number, limit = 1000) =>
    api.get<SbomDetail>(
      `/assets/${encodeURIComponent(assetId)}/sbom?tree_id=${treeId}&limit=${limit}`
    ),

  uploadSbom: async (
    assetId: string,
    treeId: number,
    file: File
  ): Promise<SbomMeta> => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await fetch(
      `/api/v1/assets/${encodeURIComponent(assetId)}/sbom?tree_id=${treeId}`,
      { method: 'POST', credentials: 'same-origin', body: formData }
    );
    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: 'Upload failed' }));
      throw new Error(
        typeof error.detail === 'string' ? error.detail : 'Upload failed'
      );
    }
    return response.json();
  },

  deleteSbom: (assetId: string, treeId: number) =>
    api.delete<void>(`/assets/${encodeURIComponent(assetId)}/sbom?tree_id=${treeId}`),
};
